### DISCUSSION TITLE
How to use rich.progress without `with` or `for` code block?

### DISCUSSION THREAD

**Discussion by yurenchen000** [Textualize/rich#1744] [Q&A] [answered]
rich.progress is very great.  thanks to developers.
// I'm not sure this is the right place to ask usage questions, sorry for bothering you

I found the offical demo, two ways usage:

A. for .. track()
```python
from rich.progress import track

for n in track(range(n), description="Processing..."):
    do_work(n)
```

B. with Progress()
```python
import time

from rich.progress import Progress

with Progress() as progress:

    task1 = progress.add_task("[red]Downloading...", total=1000)
    task2 = progress.add_task("[green]Processing...", total=1000)
    task3 = progress.add_task("[cyan]Cooking...", total=1000)

    while not progress.finished:
        progress.update(task1, advance=0.5)
        progress.update(task2, advance=0.3)
        progress.update(task3, advance=0.9)
        time.sleep(0.02)
```

<br>

they all need one code indent level for `rich.progress`,  
I just want to add progress as some log code (line by line), not change code block indent.  
(and create,update,destory may in different function)

Is that possible?

<br><br><br>

I current use trick with `progress.__enter__,  __exit__`,
this seems not the right way.
// ctrl+c not work, so I have to play with signal: `signal.signal(signal.SIGINT, handle_sigint)`
<br>

```py
import time

from rich.progress import Progress

# with Progress() as progress:

# 免 with 的调用方法
progress = Progress()
progress.__enter__()

print('hello1')
task1 = progress.add_task("[red]Downloading...1111", total=80)
task2 = progress.add_task("[green]Processing...", total=150)
task3 = progress.add_task("[cyan]Cooking...", total=100)

print('hello2')
i = 0
while not progress.finished:
    progress.update(task1, advance=0.5)
    progress.update(task2, advance=0.8)
    progress.update(task3, advance=0.9)
    i+=1
    print(f'hello {i}')
    time.sleep(0.02)

progress.__exit__(None, None, None)
```
----

**Comment by willmcgugan** [comment, marked as answer]
You can call `progress.start()` and `progress.stop()`. But I wouldn't advise it. If your code throws an exception you must ensure that `stop()` is called.
